File: frontend/src/types/userProfile.ts
import type { User } from "./models";

export interface UserServiceLink {
  id: string;
  name: string;
  status: string;
  is_active: boolean;
  category_name: string | null;
}

export interface UserHardwareAssetLink {
  id: string;
  model_name: string;
  serial_number: string | null;
  hardware_type: import("./models").HardwareType;
  quantity: number;
  status: string;
  is_active: boolean;
  hardware_location_name: string | null;
}

export interface UserProfile {
  user: User;
  owned_services: UserServiceLink[];
  assigned_services: UserServiceLink[];
  assigned_hardware_assets: UserHardwareAssetLink[];
}

export interface UserDetailOutletContext {
  profile: UserProfile;
}

File: backend/app/schemas/user_profile.py
from __future__ import annotations

import uuid

from pydantic import BaseModel, computed_field
from app.schemas.hardware import HardwareType

from app.schemas.user import UserRead


class UserServiceLinkRead(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    is_active: bool
    category_name: str | None = None


class UserHardwareAssetLinkRead(BaseModel):
    id: uuid.UUID
    model_name: str
    hardware_type: HardwareType = "laptop"
    quantity: int = 1
    serial_number: str | None
    status: str
    is_active: bool
    hardware_location_name: str | None = None


class UserProfileRead(BaseModel):
    user: UserRead
    owned_services: list[UserServiceLinkRead]
    assigned_services: list[UserServiceLinkRead]
    assigned_hardware_assets: list[UserHardwareAssetLinkRead]

    @computed_field(deprecated="Use assigned_hardware_assets")
    @property
    def assigned_laptops(self) -> list[UserHardwareAssetLinkRead]:
        return self.assigned_hardware_assets
